File: Android/TUIKit/TUICallKit/tuicallkit/src/main/java/com/tencent/qcloud/tuikit/tuicallkit/service/TUICallService.java
package com.tencent.qcloud.tuikit.tuicallkit.service;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.os.IBinder;

import androidx.core.app.NotificationCompat;

import com.tencent.qcloud.tuikit.tuicallkit.utils.DeviceUtils;

/**
 * TUICallKit are kept alive to ensure that applications are not killed by the system in the background.
 * You can delete them if they are not needed.
 */
public class TUICallService extends Service {
    private static final int NOTIFICATION_ID = 1001;

    @Override
    public void onCreate() {
        super.onCreate();
        // 获取服务通知
        Notification notification = createForegroundNotification();
        //将服务置于启动状态 ,NOTIFICATION_ID指的是创建的通知的ID
        startForeground(NOTIFICATION_ID, notification);
    }

    public static void start(Context context) {
        if (DeviceUtils.isServiceRunning(context, TUICallService.class.getName())) {
            return;
        }
        Intent starter = new Intent(context, TUICallService.class);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            context.startForegroundService(starter);
        } else {
            context.startService(starter);
        }
    }

    public static void stop(Context context) {
        Intent intent = new Intent(context, TUICallService.class);
        context.stopService(intent);
    }

    private Notification createForegroundNotification() {
        NotificationManager notificationManager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
        String notificationChannelId = "notification_channel_id_01";
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            String channelName = "TRTC Foreground Service Notification";
            NotificationChannel notificationChannel = new NotificationChannel(
                    notificationChannelId, channelName, NotificationManager.IMPORTANCE_LOW);
            notificationChannel.setDescription("Channel description");
            if (notificationManager != null) {
                notificationManager.createNotificationChannel(notificationChannel);
            }
        }
        NotificationCompat.Builder builder = new NotificationCompat.Builder(this, notificationChannelId);
        return builder.build();
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        return START_NOT_STICKY;
    }

    @Override
    public IBinder onBind(Intent intent) {
        throw new UnsupportedOperationException("Not yet implemented");
    }

    @Override
    public void onTaskRemoved(Intent rootIntent) {
        super.onTaskRemoved(rootIntent);
        stopSelf();
    }
}

File: Android/TUIKit/TUICallKit/tuicallkit-kt/src/main/java/com/tencent/qcloud/tuikit/tuicallkit/extensions/TUICallService.kt
package com.tencent.qcloud.tuikit.tuicallkit.extensions

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.Service
import android.content.Context
import android.content.Intent
import android.os.Build
import android.os.IBinder
import androidx.core.app.NotificationCompat
import com.tencent.qcloud.tuikit.tuicallkit.utils.DeviceUtils

/**
 * TUICallKit are kept alive to ensure that applications are not killed by the system in the background.
 * You can delete them if they are not needed.
 */
class TUICallService : Service() {

    companion object {
        private const val NOTIFICATION_ID = 1001
        fun start(context: Context) {
            if (DeviceUtils.isServiceRunning(context, TUICallService::class.java.name)) {
                return
            }
            val starter = Intent(context, TUICallService::class.java)
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                context.startForegroundService(starter)
            } else {
                context.startService(starter)
            }
        }

        fun stop(context: Context) {
            val intent = Intent(context, TUICallService::class.java)
            context.stopService(intent)
        }
    }

    override fun onCreate() {
        super.onCreate()
        val notification = createForegroundNotification()
        startForeground(NOTIFICATION_ID, notification)
    }

    private fun createForegroundNotification(): Notification {
        val notificationManager = getSystemService(NOTIFICATION_SERVICE) as NotificationManager
        val notificationChannelId = "notification_channel_id_01"
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val channelName = "TRTC Foreground Service Notification"
            val notificationChannel = NotificationChannel(
                notificationChannelId, channelName, NotificationManager.IMPORTANCE_LOW
            )
            notificationChannel.description = "Channel description"
            notificationManager?.createNotificationChannel(notificationChannel)
        }
        val builder = NotificationCompat.Builder(this, notificationChannelId)
        return builder.build()
    }

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        return START_NOT_STICKY
    }

    override fun onBind(intent: Intent): IBinder {
        throw UnsupportedOperationException("Not yet implemented")
    }

    override fun onTaskRemoved(rootIntent: Intent) {
        super.onTaskRemoved(rootIntent)
        stopSelf()
    }
}